BM-800: Fulfill after lock expire fixes. Change order table to support multiple fulfillment methods.

willpote opened the pull request:
This PR firstly fixes some issues with the initial implementation for fulfilling requests where the lock has expired. The first version had two functional issues:

1. It attempted to fulfill the requests immediately when the `RequestLocked` event was detected, rather than waiting until the lock had expired. This has been resolved by storing a "target_timestamp" value that represents when the lock expires, and only attempting to fulfill after that point, similar to our logic for locking requests.

2. It attempted to call the contract to fulfill lock-expired requests without pricing them. This is resolved by using `priceAndFulfillBatch` when our batch contains lock expired orders.

These issues were missed due to us not having an E2E test. I've now added an E2E test that covers the full workflow.

---

Secondly this PR makes some changes to the orders DB table aimed at reducing complexity, race conditions, and setting us up for future features like "fulfilling orders without locking" and "multiple requests with same id". 

The initial implementation monitored for `RequestLocked` events, and if detected, updated the order's OrderStatus back to Pricing, to indicate that we should reconsider the request for fulfilling after the lock expired. This introduced a couple of issues:

1. Race Conditions: for example, if we updated the status back to Pricing while the order was in the middle of proving, when the proving workflow completed the status would be overwritten to PendingAggregation, meaning the request is never re-priced.

2. Breaks the "assembly line" mental model, adding complexity. The proving node originally processed orders like an assembly line; market_monitor -> order_picker -> order_monitor -> proving -> aggregation -> submission. Allowing an orders status to be set to a previous status, breaks the simple pipeline model and complicates the state transition graph, making things harder to debug and increasing risks of bugs.

To address this, I have changed the orders table primary key to be of type `{request_id}-{request_hash}-{fulfillment_type}`. Where fulfillment type indicates whether we are intending to "Lock and Fulfill" the request, or "Wait until the lock expires and fulfill" or "Fulfill without locking the request". So now each request will result in multiple rows in the orders table, where each row represents a proof request +and+ a specific way to fulfill it, and each row goes through the assembly line independently. (Note request hash is included in the id to enable us to support multiple requests with the same id in the future.)

---

Thirdly, this PR moves capacity tracking from the order-picker into the order-monitor. Previously we would assign capacity during order-picking, which in many cases in significantly before the time that we actually need the capacity- as we must wait until the order is profitable enough to lock, or wait until the lock expires to attempt to fulfill for slashed stake. Now we only assign capacity when we hit the `target_timestamp` and are about to either lock or prove the order (depending on the fulfillment type)


---

A couple follow ups for the task will be:

Capacity limits for executor runs (this PR removes any limitations at that stage)
Executor/proving caching - this design results in operations like executor being repeated across each order row. We can cache the results
Adding a submitMerkleAndPriceAndFulfill method to the contract. Currently fulfilling both locked and unlocked orders in one transaction is not possible.

austinabell commented on crates/broker/src/lib.rs:
```diff
@@ -203,8 +213,25 @@ struct Order {
     error_msg: Option<String>,
 }
 
+// An Order represents a proof request and a specific method of fulfillment.
+//
+// Requests can be fulfilled in multiple ways, for example by locking then fulfilling them,
+// by waiting for an existing lock to expire then fulfilling for slashed stake, or by fulfilling
+// without locking at all.
+//
+// For a given request, each type of fulfillment results in a separate Order being created, with different
+// FulfillmentType values.
+//
+// Additionally, there may be multiple requests with the same request_id, but different ProofRequest
+// details. Those also result in separate Order objects being created.
+//
+// See the id() method for more details on how Orders are identified.
```
nit: This likely should be on the `Order` struct and be rustdoc comments (`///`)

austinabell commented on crates/broker/src/lib.rs:
```diff
@@ -217,14 +244,34 @@ impl Order {
         }
     }
+
+    // An Order is identified by the request_id, the fulfillment type, and the hash of the proof request.
+    // This structure supports multiple different ProofRequests with the same request_id, and different
+    // fulfillment types.
```
```suggestion
    /// An Order is identified by the request_id, the fulfillment type, and the hash of the proof request.
    /// This structure supports multiple different ProofRequests with the same request_id, and different
    /// fulfillment types.
```

austinabell commented on crates/broker/src/lib.rs:
```diff
@@ -217,14 +244,34 @@ impl Order {
+        format!(
+            "0x{:x}-{}-{:?}",
+            self.request.id,
+            self.request.signing_hash(Address::ZERO, 1).unwrap(),
```
why is this hash necessary?

willpote replied:
This is for future support of multiple requests with the same ID https://github.com/boundless-xyz/boundless/issues/162

willpote replied:
Thats right, it has been useful to have it in the id for debugging. 

Good callout on the chain and contract- I meant to come back to this actually but forgot!

And correct on fulfillment type. I think having it operate separately keeps things a lot cleaner and easier to debug, and reduces risks of various race conditions.

Regarding proof work in progress, I think we should add a caching layer in the prover to the executor runs and proof runs (I think this is something we should do regardless of this PR anyway for DDOS/spam protection)

austinabell commented on crates/broker/src/aggregator.rs:
```diff
@@ -159,21 +159,21 @@ impl AggregatorService {
     }
 
-    async fn prove_assessor(&self, order_ids: &[U256]) -> Result<String> {
+    async fn prove_assessor(&self, order_ids: Vec<String>) -> Result<String> {
```
this doesn't seem to need to clone the vec of orders? It just uses a ref to the items?
```suggestion
    async fn prove_assessor(&self, order_ids: &[String]) -> Result<String> {
```

austinabell reviewed (commented):
Partial review because still personally unclear the motivation and if this is a good long-term option. Primarily:

> 1. Race Conditions: for example, if we updated the status back to Pricing while the order was in the middle of proving, when the proving workflow completed the status would be overwritten to PendingAggregation, meaning the request is never re-priced.
> 2. Breaks the "assembly line" mental model, adding complexity. The proving node originally processed orders like an assembly line; market_monitor -> order_picker -> order_monitor -> proving -> aggregation -> submission. Allowing an orders status to be set to a previous status, breaks the simple pipeline model and complicates the state transition graph, making things harder to debug and increasing risks of bugs.
> 
> To address this, I have changed the orders table primary key to be of type `{request_id}-{request_hash}-{fulfillment_type}`. Where fulfillment type indicates whether we are intending to "Lock and Fulfill" the request, or "Wait until the lock expires and fulfill" or "Fulfill without locking the request". So now each request will result in multiple rows in the orders table, where each row represents a proof request +and+ a specific way to fulfill it, and each row goes through the assembly line independently. (Note request hash is included in the id to enable us to support multiple requests with the same id in the future.)

The pattern that the broker tries to fulfill two orders at the same time feels more like a bug than something to handle with more complexity. Wondering if this could be solved with a compare and swap pattern on updates rather than handling in parallel.

willpote commented:
> Partial review because still personally unclear the motivation and if this is a good long-term option. Primarily:
> 
> > 1. Race Conditions: for example, if we updated the status back to Pricing while the order was in the middle of proving, when the proving workflow completed the status would be overwritten to PendingAggregation, meaning the request is never re-priced.
> > 2. Breaks the "assembly line" mental model, adding complexity. The proving node originally processed orders like an assembly line; market_monitor -> order_picker -> order_monitor -> proving -> aggregation -> submission. Allowing an orders status to be set to a previous status, breaks the simple pipeline model and complicates the state transition graph, making things harder to debug and increasing risks of bugs.
> > 
> > To address this, I have changed the orders table primary key to be of type `{request_id}-{request_hash}-{fulfillment_type}`. Where fulfillment type indicates whether we are intending to "Lock and Fulfill" the request, or "Wait until the lock expires and fulfill" or "Fulfill without locking the request". So now each request will result in multiple rows in the orders table, where each row represents a proof request +and+ a specific way to fulfill it, and each row goes through the assembly line independently. (Note request hash is included in the id to enable us to support multiple requests with the same id in the future.)
> 
> The pattern that the broker tries to fulfill two orders at the same time feels more like a bug than something to handle with more complexity. Wondering if this could be solved with a compare and swap pattern on updates rather than handling in parallel.

Yeah I considered something like that, but I actually felt like this approach would lead to less complexity. With more of a compare/swap every DB operation need to run a check, we have some optional fields that are populated later in the flow that we'd have to make sure to clear to maintain data integrity, and I think its harder to debug if we're moving backwards and forwards through Order Statuses. 

Also note, the orders don't fully end up being handled in parallel. More that we end up cancelling (setting to Skip) the existing order, and kick off a new order where we try to fulfill after the lock expires.

willpote commented on crates/broker/src/lib.rs:
```diff
@@ -140,8 +140,10 @@ enum OrderStatus {
     Pricing,
-    /// Order is ready to lock at target_timestamp
-    Locking,
+    /// Order is ready to lock at target_timestamp and then be fulfilled
+    FulfillAfterLocking,
+    /// Order is ready to be fulfilled when its lock expires at target_timestamp
+    FulfillAfterLockExpire,
```
Yes, definitely a breaking change, but I agree I think its ok for testnet

nategraf commented on crates/broker/src/market_monitor.rs:
```diff
@@ -282,14 +317,24 @@ where
-                        if let Err(e) = db.set_order_complete(U256::from(event.requestId)).await {
-                            tracing::error!("Failed to update order status to Done: {e:?}");
+                        if let Err(e) = db
+                            .set_request_fulfilled(
```
Do we need to distinguish here on whether or not the prover was paid? Two cases where this distinction may matter if when someone fulfills an order we already locked (i.e. they don't get paid, and we are still eligible for payment), and if the order is not locked and someone else fills it but doesn't get paid due to client not having enough funds (https://github.com/boundless-xyz/boundless/issues/545). The latter of the two edge cases may be niche enough to ignore.

willpote replied:
That would be useful to store

I'm inclined to leave this to a follow up as 1/ Currently the last point at which we consider aborting a fulfillment is right before we lock it, so handling these cases would be new code. and 2/ Its actually quite tricky to get this info at the moment since the payment failed event doesn't include the request id

nategraf reviewed (commented):
Left some comments. I like the idea of having a table for events like lock and fulfill that can allow us to determine the onchain state of an order by ID, separate from the order indexed by request digest. Not sure about whether this should be one table with state or two like it is now. I think both work though.

willpote commented on crates/broker/src/lib.rs:
```diff
@@ -140,8 +140,10 @@ enum OrderStatus {
     Pricing,
-    /// Order is ready to lock at target_timestamp
-    Locking,
+    /// Order is ready to lock at target_timestamp and then be fulfilled
+    FulfillAfterLocking,
+    /// Order is ready to be fulfilled when its lock expires at target_timestamp
+    FulfillAfterLockExpire,
```
> It unclear to me why these are two different states vs being a case where the the state of the order may skip Locking and go straight to PendingProving or Proving when we don't intend to lock it.

`PendingProving` schedules it to be proved immediately, but in this case we want to wait until the lock expires (to see if the locker actually fulfills it or not) before proving.

I agree with your concerns on the state machine, though I actually think as part of Broker RPC separation we should be moving towards a different architecture for the DB- I've always found designs where multiple services are updating the same DB to be error prone and hard to scale, and I've seen many people consider it a bit of an anti-pattern ([e.g.](https://martinfowler.com/bliki/IntegrationDatabase.html)). 

My take is we should move more to a design where each service owns its own data store, and we strive to make the rows in each immutable

willpote replied:
For the names open to ideas. Maybe: `WaitingToLock` and `WaitingForLockExpire`? I think its useful to align the naming to indicate that these are both states where we are waiting until some point in the future

willpote commented on crates/broker/src/order_monitor.rs:
```diff
@@ -211,42 +299,411 @@ where
+                tracing::debug!("Order {} estimated to take {} seconds, and would be completed at {}. It expires at {}", order.id(), proof_time_seconds, completion_time, expiration);
+
+                if completion_time > expiration {
+                    tracing::info!("Order {:x} cannot be completed before its expiration at {}, proof estimated to complete at {}. Skipping", order.request.id, expiration, completion_time);
```
Added it back

willpote commented on crates/broker/src/order_monitor.rs:
```diff
@@ -211,42 +299,411 @@ where
+
+            let proof_time_seconds = total_commited_cycles.div_ceil(1_000).div_ceil(peak_prove_khz);
+            let mut prover_available_at = started_proving_at + proof_time_seconds;
+            tracing::debug!("Already committed to {} orders, with a total cycle count of {}, a peak khz limit of {}, started working on them at {}, we estimate the prover will be available at {}", num_commited_orders, total_commited_cycles, peak_prove_khz, started_proving_at, prover_available_at);
```
Added a condition to skip when 0 orders

austinabell commented on crates/broker/src/order_monitor.rs:
```diff
@@ -211,42 +299,411 @@ where
+            final_orders.len()
+        );
+        tracing::debug!(
+            "Final orders ready for locking and/or proving after filtering: {final_orders:?}"
```
sorry I mixed up logs from https://github.com/boundless-xyz/boundless/pull/542/files/574d87e616bbb2be90aacc27194db23cc7fdabfa#r2066818924 this was the one I meant to point to. Perhaps if info can be useful it's made as a trace log for our debugging purposes? Debug is on by default with our default broker setup right now

willpote replied:
Oh, so I implemented a custom debug for this that keeps it short: https://github.com/boundless-xyz/boundless/pull/542/files#diff-99ebc229b742243f5d295e008af90162c013ee6455ac3d0c49f3f86ef9c2295eR84-R95

I've found it quite useful to have on debug mode, but happy to move to trace if you feel strongly

austinabell commented on crates/broker/src/order_monitor.rs:
```diff
@@ -211,42 +299,411 @@ where
+        }
+
+        tracing::info!(
+            "Started with {} orders ready to be locked and/or proven. After applying capacity limits of {} max concurrent proofs and {} peak khz, filtered to {} orders: {:?}",
```
https://github.com/boundless-xyz/boundless/pull/542/files/574d87e616bbb2be90aacc27194db23cc7fdabfa#r2066906879 mixed up the logs, sorry
